src/fixate/core/switching.py

```diff
@@ -0,0 +1,743 @@
+    ) -> None:
+        raise NotImplementedError
+
+    # not used in any scripts
+    def update_pins_by_name(
+        self, pins: Collection[Pin], trigger_update: bool = True
+    ) -> None:
+        raise NotImplementedError
+
+    def __getitem__(self, item: Pin) -> bool:
+        """Get the value of a pin. (only inputs? or state of outputs also?)"""
+        raise NotImplementedError
+
+    def __setitem__(self, key: Pin, value: bool) -> None:
+        """Set a pin"""
+        raise NotImplementedError
+
+
+class MuxGroup:
+    """
+    Group multiple VirtualMux's, for use in a single Jig Driver.
+
+    If a test script, it is expected that MuxGroup will be subclassed, with attributes
+    being each required VirtualMux subclass. This can be done using a dataclass:
+
+    @dataclass
+    class JigMuxGroup(MuxGroup):
+        mux_one: MuxOne = field(default_factory=MuxOne)
+        mux_two: MuxTwo = field(default_factory=MuxTwo)
+    """
+
+    def get_multiplexers(self) -> list[VirtualMux]:
+        return [attr for attr in self.__dict__.values() if isinstance(attr, VirtualMux)]
+
+    def reset(self) -> None:
+        mux_list = self.get_multiplexers()
+        if len(mux_list) == 0:
+            return
+
+        for mux in mux_list[:-1]:
+            mux.reset(trigger_update=False)
+        mux_list[-1].reset(trigger_update=True)
+
+    def active_signals(self) -> list[str]:
+        return [str(mux) for mux in self.get_multiplexers()]
+
+
+JigSpecificMuxGroup = TypeVar("JigSpecificMuxGroup", bound=MuxGroup)
+
+
+class JigDriver(Generic[JigSpecificMuxGroup]):
+    """
+    Combine multiple VirtualMux's and multiple AddressHandler's.
+
+    The jig driver joins muxes to handlers by matching up pin definitions.
+    """
+
+    def __init__(
+        self,
+        mux_group_factory: Callable[[], JigSpecificMuxGroup],
+        handlers: Sequence[AddressHandler],
+    ):
+        # keep a reference to handlers so that we can close them if required.
+        self._handlers = handlers
+        self.virtual_map = VirtualAddressMap(handlers)
+
+        self.mux = mux_group_factory()
+        for mux in self.mux.get_multiplexers():
+            # Perhaps we should instantiate the virtual mux here
+            # and pass in the virtual_map.add_update. But we'd have to do some
+            # magic in the MuxGroup call to pass add_update to each VirtualMux
+            # constructor, and I was hoping to just use a dataclass...
+            mux._update_pins = self.virtual_map.add_update
+
+    def __setitem__(self, key: Pin, value: bool) -> None:
+        self.virtual_map.update_pin_by_name(key, value)
+
+    def __getitem__(self, item: Pin) -> bool:
+        return self.virtual_map[item]
+
+    def close(self) -> None:
+        for handler in self._handlers:
+            handler.close()
+
+    def active_pins(self) -> Set[Pin]:
+        return self.virtual_map.active_pins()
+
+    def reset(self) -> None:
+        """
```

Comment:
I'm glad you ask, because I'm not sure and it's worth discussing.

There was a mess before of "default signals" on muxs, "default pins" on address handlers and a mux would implicitly handle an empty string signal as "no pins". `JigDriver.reset()` would set default on address handlers *and* set each mux to it's default signal.

```python
VirtualMux.reset()  # set this mux to signal ""
MuxGroup.reset()  # set all muxes to signal ""
VirtualAddressMap.reset()  # set all pins to off
JigDriver.reset()  # just calls VirtualAddressMap.reset() 
```

My thinking was that `jig.reset()` would be used to reset all pins and `jig.mux.reset()` would be used to reset all muxes. Those two should be equivalent, if every pin is used in a mux. Having said that, I don't know how useful the distinction is and if people writing script would be able to keep the subtle difference in their heads

Perhaps a more useful approach is
```python
jig.reset()  # reset all muxes to signal "" (via the mux group reset method)

# then if a script really needs it
jig.virtual_map.reset()  # would reset all pins, without changing mux states.
```

Reply:
TBH, I think once all jig scripts were switched over and `jig.reset()` was changed to `jig.mux.reset()` people would copy and paste a similar script and follow the script (pun intended), and not even think about/know about/remember the difference, until someone thinks of a reason to treat the pins and mux states differently and goes looking for a way to do it.

Though I don't mind the specificity of `jig.mux.reset()`, because most of the time that call is made it's with the intent to putting the muxes to their off state.

Do you have any thoughts on this @daniel-montanari ?

Reply:
Maybe we could remove `jig.reset()`? Then someone writing the script has to be more conscious about choosing `jig.mux.reset()` or `jig.virtual_map.reset()`. I don't actually like that idea...

I'll wait and see what Dan has to say!

Reply:
I would think of resetting the jig more like resetting the muxes (which turns all the pins off). Does this mean it turns the pins off but leaves the muxes in their current state? What happens the next time I set the state of a mux and update the pins?

I'm imagining being a grad, resetting the jig and then probing away at stuff and turning one of the muxes back on to check a connection. I'd want the actual behaviour to match the expected behaviour.

Reply:
> What happens the next time I set the state of a mux and update the pins?
If you update a mux, it will write update the value of all of its pins (but will not change pins from any other mux)

I think I'm coming back to what I wrote a few comments up
```python
jig.reset()  # reset all muxes to signal "" (via the mux group reset method)

# then if a script really needs it
jig.virtual_map.reset()  # would reset all pins, without changing mux states.
```

> I'd want the actual behaviour to match the expected behaviour.
I just want everyone to agree on the expected behaviour :)

Reply:
I'm happy to adopt that approach.

Reply:
Done